opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -95,14 +95,140 @@ message InstrumentationLibraryMetrics {
+//
+// ## StartTimeUnixNano: Gauges
+//
+// The use of StartTimeUnixNano is optional, recommended for overlap
+// protection and to express explicitly absent points.  When used,
+// StartTimeUnixNano should be set as for either cumulative or delta
+// aggregation temporality, meaning the start timestamp can be set to
+// the previous point's `StartTimeUnixNano` or the previous point's
+// `TimeUnixNano` to encode an unbroken sequence of Gauge observations.
+//
+// ## StartTimeUnixNano: Summaries
+//
+// Summary points are treated the same as for cumulative aggregation
+// temporality, by convention.  StartTimeUnixNano should be set to the
+// starting timestamp in an unbroken sequence of observations.
+//
+// # Unknown StartTimeUnixNano
+//
+// This protocol adopts the Prometheus heuristic for reset detection,
+// and with it prescribes a solution for filling in StartTimeUnixNanos
+// with as much information as possible to assist consumers with
+// calculating rates and detecting overlap.  This process can be
+// applied to Sum and Histogram points with cumulative aggregation
+// temporality, Gauge, and Summary points.
+//
+// The Prometheus heuristic indicates that a reset must have occured
+// in a monotonic series whenever the value descends from one value to
+// the next, in a timestamp-ordered sequence.
+//
+// To facilitate reset and overlap detection, when single-timestamped
+// sample values are translated into OTLP, such as from the Prometheus
+// Remote-Write protocol, use the following.
+//
+// ## Filling in StartTimeUnixNano
+//
+// Note: This is meant to provide guidance in importing data into OTLP
+// that does not have start time and there is no other form of reset
+// detection.
+//
+// To import timestamped cumulative measurements into into OTLP using
+// a stateful processor, the input must be processed in order.  A
+// series of points that formed the beginning of the sequence,
+// including an unknown number of resets, are assumed to be lost or
+// previously recorded when the processor begins observing points.
+// Then:
+//
+// 1. Keep a map, using the metric name, resource, and attributes as
+//    the key, containing a struct of three parts:
+//    - resetTimestamp: the first observation or the reset timestamp
+//    - previousTimestamp: the timestamp of the most recent observation
+//    - previousValue: a monotonic value
+// 2. When a new point arrives and it is the first observation of that particular key:
+//    - in the map, set resetTimestamp to the arriving timestamp
+//    - in the map, set previousTimestamp to the arriving timestamp
+//    - in the map, set previousValue to the arriving value
+//    - in the output, set StartTimeUnixNano to the (new) resetTimestamp
+// 3. For each subsequent point that arrives in the sequence:
+//    - if the sample value is less than previousValue,
+//      then set resetTimestamp to previousTimestamp
+//    - set previousValue to the arriving value
```

Comment:
Thank you @jmacd for detailed comments. For 3.2, if we set the previous value to current value (when not detecting resets) then this won't be cumulative (since we are not subtracting from first value when resets are not detected) ?. Also with this approach, the monotonically increasing metric values would still be different for ex;- between a prometheus counter collected thru Remote write v. collected thru OTEL collector. May be i am missing something, apologies if thats the case. Perhaps we need to add up when exporting is what it is ?

Reply:
I may have mis-understood you, but I think the following is not true:

> the monotonically increasing metric values would still be different for ex;- between a prometheus counter collected thru Remote write v. collected thru OTEL collector.

Prometheus Remote Write writes the reset value, and the algorithm documented here just helps us the observations in a way that overlapping data can be detected. The proposal here is made so that PRW converted to OTLP converted back to PRW does not alter the data.

There has been a discussion about re-aggregation, mentioned as _another_ valid way to translate PRW into OTLP, which is to insert a constant negative offset so the series has a "true reset". See https://github.com/open-telemetry/opentelemetry-specification/pull/1646

I spec'd out re-aggregation further in https://github.com/open-telemetry/opentelemetry-specification/pull/1649, where there is an explicit goal of being able to do what you're asking for. The idea that resets are erased from the stream and cumulative values continue to rise can be treated as a general case of reaggregation, as follows:
(1) Apply the conversion described here, so that resets are explicitly encoded in the stream
(2) Apply a cumulative-to-delta translation
(3) Apply a delta-to-cumulative translation again.

Usually this process is done in connection with temporal alignment (see https://github.com/open-telemetry/opentelemetry-specification/pull/1648) and the cumulative value is calculated to be zero at a specific point (usually the oldest timestamp in the query).